### ISSUE TITLE
Static Quantized model accuracy varies greatly with Calibration data

### ISSUE THREAD

**Issue by neeraj-j** [pytorch/pytorch#45185]
## 🐛 Bug

Changing the calibration data impacts the generated quantized model's accuracy massively. On larger dataset (COCO) and complex models the validation accuracy varies from 0 - 87.3% (actual 90.6). 

## To Reproduce

1. Run the Colab notebook given at "https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html"
2.  You will have to add following cell to download data and model 
```
import requests
import zipfile
 
url = 'https://s3.amazonaws.com/pytorch-tutorial-assets/imagenet_1k.zip'
filename = './imagenet_1k_data.zip'
r = requests.get(url)
 with open(filename, 'wb') as f:
    f.write(r.content)
zip = zipfile.ZipFile(filename)
zip.extractall()
#download model
url = 'https://download.pytorch.org/models/mobilenet_v2-b0353104.pth'
filename = 'mobilenet_pretrained_float.pth'
r = requests.get(url)
with open(filename, 'wb') as f:
    f.write(r.content)
```
3. Run the following cell in loop
```
for i in range(10):    # <----- Added loop
  per_channel_quantized_model = load_model(saved_model_dir + float_model_file)
  per_channel_quantized_model.eval()
  per_channel_quantized_model.fuse_model()
  per_channel_quantized_model.qconfig = torch.quantization.get_default_qconfig('fbgemm')
  #print(per_channel_quantized_model.qconfig)

  torch.quantization.prepare(per_channel_quantized_model, inplace=True)
  evaluate(per_channel_quantized_model,criterion, data_loader, num_calibration_batches)
  torch.quantization.convert(per_channel_quantized_model, inplace=True)
  top1, top5 = evaluate(per_channel_quantized_model, criterion, data_loader_test, neval_batches=num_eval_batches)
  print('Evaluation accuracy on %d images, %2.2f'%(num_eval_batches * eval_batch_size, top1.avg))
torch.jit.save(torch.jit.script(per_channel_quantized_model), saved_model_dir + scripted_quantized_model_file)
```
4. You might see accuracy varies from 73.33% - 78.0% 
5. Now change test_sampler = torch.utils.data.SequntialSampler(dataset_test) to RandomSampler()
6. Again run the Notebook. You will see accuracy variation from 66.0 - 76%. 

That's 10% variation on such a small dataset and simple model. On Larger dataset like COCO and complex model, I see variation from 0-87.3% (actual 90.6%). This uncertainty makes PyTorch quantization unsuitable for production purpose . There is no way to predict how to get best accuracy. 

Even Quantized aware training has given me best accuracy of 80%. 

## Expected behavior

Static-Quantization process should generate model with deterministic accuracy.

## Environment

 - PyTorch Version (e.g., 1.0): 1.6.0+cu101
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): Colab (default version)
 - Build command you used (if compiling from source): NA
 - Python version: 1.7
 - CUDA/cuDNN version: NA
 - GPU models and configuration: NA
 - Any other relevant information:

## Additional context

cc @jerryzh168 @jianyuh @dzhulgakov @raghuramank100 @jamesr66a @vkuzo

**Comment by vkuzo** [comment]
hi @neeraj-j , thanks for the report.  It's likely that the numbers you are seeing can be explained by the dataset composition.  The dataset referenced in the tutorial is not the real ImageNet1K dataset, it is a tiny subsample of it with only 1k images.  

To get consistent measurements on this toy dataset, you could do the following:
* always evaluate on the same slice of the dataset (i.e. create the eval dataset before your for loop)
* after your split calibrate vs eval, try calibrating on different slices of calibrate, and verify that you get similar accuracy scores on the same eval slice

Using a real dataset would also help get consistent measurements.  We will add a warning to the tutorial to clarify that the dataset used is a toy one.

**Comment by neeraj-j** [comment]
Hi @vkuzo ,  I understand your response. 
1. As mentioned in point # 4 above, post-quantization eval dataset slice is same (sequential sampler) still I see variation of ~5%. Which is huge for such tiny dataset. Below is the output with same slice of eval dataset ( I printed target values to make sure that every loop is evaluated on same data slice 0-299) by adding "if prnt : print(target)"

```
........../usr/local/lib/python3.6/dist-packages/torch/quantization/observer.py:877: UserWarning: must run observer before calling calculate_qparams.                                    Returning default scale and zero point 
  Returning default scale and zero point "
....................Evaluation accuracy on 300 images, 77.33
....................Evaluation accuracy on 300 images, 74.67
....................Evaluation accuracy on 300 images, 75.33
....................Evaluation accuracy on 300 images, 78.00
....................Evaluation accuracy on 300 images, 76.67
....................Evaluation accuracy on 300 images, 74.67
....................Evaluation accuracy on 300 images, 76.67
....................Evaluation accuracy on 300 images, 76.00
....................Evaluation accuracy on 300 images, 76.67
....................Evaluation accuracy on 300 images, 73.33
```
2. For larger dataset like COCO, the issue does not get resolved, It gets aggravated. That's my concern here. I am seeing variation of 0-87% on COCO Dataset. I am doing post quantization eval on full val dataset for coco (not partial)
3. How can I get consistent best result on large dataset.

**Comment by vkuzo** [comment]
Two important things to consider would be to (1) make sure the calibration data represents your true population and (2) the evaluation data represents your true population.

For the toy dataset, both (1) and (2) would be challenging as the dataset is tiny and it not supposed to be a real dataset.

For a real dataset, I'd recommend increasing the size of both your calibration and evaluation slices to make them more representative of the actual population.  300 images seems like it might not be a large enough slice.  Do you see more consistent results if you increase the size of eval and calibration slices?

**Comment by neeraj-j** [comment]
@vkzo, It will be helpful if you read my responses. 

I am working on this issue for 2 months now and I have tried all the things that you are suggesting. 

Can you share any object detection quantized model trained on "large data".

**Comment by vkuzo** [comment]
When you see this behavior in experiments non-toy datasets, what is the size of your evaluation slice, calibration slice, total dataset?

**Comment by neeraj-j** [comment]
I have already mentioned above that I am using COCO dataset. It has more that 125,000 calibration slice and 30,000 eval slice. 

I have also mentioned that as I increase calibration slice the accuracy keep on decreasing till it settles at 79%(quantized) from 91.6%(float model).

**Comment by vkuzo** [comment]
hi @neeraj-j , thanks for the additional information.  91.6% fp accuracy to 79% static quantization accuracy indeed sounds like a large drop.  There are two separate things which could be useful to debug:

1. Regardless of whether the model is sensitive to quantization or not (if the drop from fp to quant accuracy is large or not), cycling between representative calibration datasets shouldn't impact eval accuracy too much.  If you see it changing significantly, it could mean that one of the slices is not representative of the other.  In you latest answer you mentioned that the quantized accuracy settles at 79% as calibration slice increases - this is evidence that the true accuracy is close to 79%.

2. If you are seeing a drop from 91.6% to 79%, the model could be very sensitive to quantization.  One avenue to debug this could be using the Numeric Suite (https://github.com/pytorch/tutorials/blob/master/prototype_source/numeric_suite_tutorial.py) to see if there is a particular layer which is problematic and has high quantization error.  The model could then be requantized again with adjusted quantization settings for the problematic layer, or skipping quantization of that layer altogether.

**Comment by neeraj-j** [comment]
Thanks @vkzo this looks like a good pointer. I will look into it in couple of days. 
In the mean while, I am wondering, if you have any quantization demo on Object detection, keypoints, image segmentation.  else I will create it on large data set and share it.

**Comment by vkuzo** [comment]
> Thanks @vkzo this looks like a good pointer. I will look into it in couple of days.
> In the mean while, I am wondering, if you have any quantization demo on Object detection, keypoints, image segmentation. else I will create it on large data set and share it.

hi @neeraj-j , check out https://github.com/pytorch/vision/blob/master/references/classification/train_quantization.py for MobileNetV2 quantization reference code.  For detection and keypoints, what people often do is quantize the trunk and leave RPN and heads in floating point, although quantizing the heads would be possible as well.  I don't believe we have a tutorial on these nor on segmentation but the basic concept would be similar. One could profile the floating point model, try quantizing the most expensive layers, and find a point on the performance-accuracy tradeoff which makes sense for the specific use case.

**Comment by neeraj-j** [comment]
Thanks, I'll get back to you..

**Comment by neeraj-j** [comment]
Using numeric suite I was able to debug and fix the issue.
Appreciate all your help and time.
